src/rules/prefer-readonly/classScope.ts

```diff
@@ -0,0 +1,127 @@
+/**
+ * @license
+ * Copyright 2017 Palantir Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as utils from "tsutils";
+import * as ts from "typescript";
+
+const OUTSIDE_CONSTRUCTOR = -1;
+const DIRECTLY_INSIDE_CONSTRUCTOR = 0;
+
+export type ParameterOrPropertyDeclaration = ts.ParameterDeclaration | ts.PropertyDeclaration;
+
+function typeIsOrHasBaseType(type: ts.Type, parentType: ts.Type) {
+    if (type.symbol === undefined || parentType.symbol === undefined) {
+        return false;
+    }
+
+    const typeAndBaseTypes = [type];
+    const ancestorTypes = type.getBaseTypes();
+
+    if (ancestorTypes !== undefined) {
+        typeAndBaseTypes.push(...ancestorTypes);
+    }
+
+    for (const baseType of typeAndBaseTypes) {
+        if (baseType.symbol !== undefined && baseType.symbol.name === parentType.symbol.name) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+export class ClassScope {
```

Comment:
I know there's one bit of `readonly`-specific logic here, but I'd slightly prefer moving this class to `src/language/classScope.ts`

I know I mentioned in another thread that it's cool to have multiple files for a rule implementation, but in this case I think we can reasonably avoid it.